server/ips.go

```diff
@@ -20,3 +20,143 @@ func GetOutboundIP() (net.IP, error) {
 
 	return localAddr.IP, nil
 }
+
+// addrToIPNet casts addr to IPNet.
+// Returns nil if addr is not of IPNet type.
+func addrToIPNet(addr net.Addr) *net.IPNet {
+	switch v := addr.(type) {
+	case *net.IPNet:
+		return v
+	default:
+		return nil
+	}
+}
+
+// FilterAddressesForPairingServer filters private unicast addresses.
+// ips is a 2-dimensional array, where each sub-array is a list of IP
+// addresses for a single network interface.
+func FilterAddressesForPairingServer(ips [][]net.IP) []net.IP {
+	var result []net.IP
+
+	for _, niIps := range ips {
+		var ipv4, ipv6 []net.IP
+
+		for _, ip := range niIps {
+
+			// Only take private global unicast addrs
+			if !ip.IsGlobalUnicast() || !ip.IsPrivate() {
+				continue
+			}
+
+			if v := ip.To4(); v != nil {
+				ipv4 = append(ipv4, ip)
+			} else {
+				ipv6 = append(ipv6, ip)
+			}
+		}
+
+		// Prefer IPv4 over IPv6 for shorter connection string
+		if len(ipv4) == 0 {
+			result = append(result, ipv6...)
+		} else {
+			result = append(result, ipv4...)
+		}
+	}
+
+	return result
+}
+
+// GetLocalAddresses returns an array of all addresses
+// of all available network interfaces.
+func GetLocalAddresses() ([][]net.IP, error) {
+	nis, err := net.Interfaces()
+	if err != nil {
+		return nil, err
+	}
+
+	var ips [][]net.IP
+
+	for _, ni := range nis {
+		var niIps []net.IP
+
+		addrs, err := ni.Addrs()
+		if err != nil {
+			return nil, err
+		}
+
+		for _, addr := range addrs {
+
+			var ip net.IP
+			if ipNet := addrToIPNet(addr); ipNet == nil {
+				continue
+			} else {
+				ip = ipNet.IP
+			}
+
+			niIps = append(niIps, ip)
+		}
+
+		if len(niIps) > 0 {
+			ips = append(ips, niIps)
+		}
+	}
+
+	return ips, nil
+}
+
+// GetLocalAddressesForPairingServer is a high-level func
+// that returns a list of addresses to be used by local pairing server.
+func GetLocalAddressesForPairingServer() ([]net.IP, error) {
+	ips, err := GetLocalAddresses()
+	if err != nil {
+		return nil, err
+	}
+	return FilterAddressesForPairingServer(ips), nil
+}
+
+// FindReachableAddresses returns a filtered remoteIps array,
+// in which each IP matches one or more of given localNets.
+func FindReachableAddresses(remoteIps []net.IP, localNets []net.IPNet) []net.IP {
+	var result []net.IP
+	for _, localNet := range localNets {
+		for _, remoteIP := range remoteIps {
+			if localNet.Contains(remoteIP) {
+				result = append(result, remoteIP)
+			}
```

Comment:
Is it possible that both devices may each have different networks but that both these networks contain a the same ip?

- Device a 
  - network a (with ip 1.2.3.4)
  - network b (with ip 5.6.7.8)
  - network c (with ip 9.0.1.2)
- Device b
  - network d (with ip 3.4.5.6)
  - network e (with ip 7.8.9.0)
  
Now imagine that network d also has an ip address `1.2.3.4` . This function would in theory give us a false positive as the devices don't share a network.

I think we need a 2nd step of attempting to call each `result` ip address, to make sure that we can communicate with the other device.

Reply:
First of all, afaik, it's an invalid network configuration when 2 devices has same IP address.
Probably we could (should?) still handle such cases and not consider same IPs as reachable. Do you think this would be good?

> I think we need a 2nd step of attempting to call each result ip address, to make sure that we can communicate with the other device.

This is why I eventually decided to return an array of IPs.
We can add another procedure to attempt to connect to each of them, but I thought maybe LP Client would do that 🤔

Reply:
> Now imagine that network d also has an ip address `1.2.3.4` . This function would in theory give us a false positive as the devices don't share a network.

Ah, I wasn't clear. "*Now imagine that network d also has an ip address `1.2.3.4`*", what I should have said is "*Now imagine that network d also has an ip address `1.2.3.4`* **belonging to a device c**"

> We can add another procedure to attempt to connect to each of them, but I thought maybe LP Client would do that 🤔

Yes doing that in another stage is fine, I remember that part of the conversation;

Reply:
Ok, I get you now 
Yes, that is possible and this algorithm will definitely take this as a reachable address. I thinkg this is a good case when UDP discovery should come into the house 🙂